File: web/src/utils/json.ts
// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export const ToString = (o: any) => {
  if (o === null || o === undefined) {
    return "";
  }
  if (typeof o == "string") {
    return o;
  } else if (Array.isArray(o)) {
    let tmp = "";
    for (let i = 0; i < o.length; i++) {
      tmp += (tmp === "" ? "" : ", ") + ToString(o[i]);
    }
    return `[${tmp}]`;
  } else if (typeof o == "object") {
    let tmp = "";
    for (const key in o) {
      tmp += (tmp === "" ? "" : ", ") + `"${key}":` + ToString(o[key]);
    }
    return `{${tmp}}`;
  } else {
    return o.toString();
  }
};

export const byString = (o: any, s: string) => {
  if (s == undefined) {
    return "";
  }
  if (s in o) {
    return ToString(o[s]);
  }
  s = s.replace(/\[(\w+)\]/g, ".$1"); // convert indexes to properties
  s = s.replace(/^\./, ""); // strip a leading dot
  const a = s.split(".");

  // eg.1, o = {"a1": {"b1": "1234", "b2": "1", "b3": {"c1":"12", "c2":"22"}}, "a2":"334"}, s = "a1.b3.c2"
  // eg.2, o = {"a1.b1": "1234", "a1":"233", "a2":"334"}, s = "a1.b1"
  // eg.3, o = {"a1.b1": {"c1": "123"}, "a1":"233", "a2":"334"}, s = "a1.b1.c1"
  // eg.4, o = {"a1": [{"b1":"12", "b2":"222"}], "a2":"334"}, s = "a1"
  for (let i = 0, n = a.length; i < n; ++i) {
    let flag = false;
    for (let j = n; j > i; j--) {
      // Priority matching (longest)
      const key = a.slice(i, j).join(".");
      if (typeof o == "object" && key in o) {
        o = o[key];
        flag = true;
        break;
      }
    }

    if (flag == false) {
      o = "";
    }
  }
  return ToString(o);
};

export const deepKeys = (o: any) => {
  if (!(o instanceof Object)) {
    return [];
  }
  const results = [];
  const keys = Object.keys(o);
  for (const i in keys) {
    // Check if we can use === instead of ==
    if (o[keys[i]] == undefined || o[keys[i]].length) {
      results.push(keys[i]);
    } else {
      const subKeys = deepKeys(o[keys[i]]);
      if (subKeys.length > 0) {
        subKeys.forEach((key) => {
          results.push(keys[i] + "." + key);
        });
      } else {
        results.push(keys[i]);
      }
    }
  }
  return results;
};

File: src/common/utils/json.rs
// Copyright 2023 Zinc Labs Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

pub use serde_json::{
    from_slice, from_str, from_value, json, to_string, to_value, to_vec, Error, Map, Number, Value,
};

pub fn estimate_json_bytes(val: &Value) -> usize {
    let mut size = 0;
    match val {
        Value::Object(map) => {
            // {?} extra 2
            size += 2;
            for (k, v) in map {
                // "key":?, extra 4 bytes
                size += k.len() + estimate_json_bytes(v) + 4;
            }
            // remove ',' for last item
            if !map.is_empty() {
                size -= 1;
            }
        }
        Value::Array(arr) => {
            // []=>2 [?]=>2 [?,?] extra 1+n
            size += std::cmp::max(arr.len(), 1) + 1;
            for v in arr {
                size += estimate_json_bytes(v);
            }
        }
        Value::String(s) => {
            // "?"=>2
            size += s.len() + 2;
        }
        Value::Number(n) => {
            size += n.to_string().len();
        }
        Value::Bool(b) => {
            // true for 4 bytes, false for 5 bytes
            size += if *b { 4 } else { 5 };
        }
        Value::Null => {
            size += 4;
        }
    }
    size
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_estimate_json_bytes() {
        let json = r#"{"a":null,"b":true,"c":false,"d":{"a":"b","c":true,"d":false,"e":123456},"e":[""],"f":["a"],"g":["a","b"],"h":"bcdef","i":{},"j":{"ok":"yes"}}"#;
        let val: Value = from_str(json).unwrap();
        assert_eq!(estimate_json_bytes(&val), json.len());
    }

}
